### PR TITLE
MM-30407: improve slow typing

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
Skip rendering the sidebar channel menu until open. This goes hand-in-hand with the [linked Redux PR](https://github.com/mattermost/mattermost-redux/pull/1407) to address significant performance issues with the (new) sidebar that ultimately affects typing performance on busy servers.

#### Ticket Link
Fixes: https://mattermost.atlassian.net/browse/MM-30407

### CODE REVIEW

**Comment by lieut-data** [comment]
(Will fix tests re: not rendering the menu when not open.)

**Comment by hmhealey** [comment]
Ah, the menu component. We meet again. This sounds good to me. I'd be slightly worried about this breaking any opening/closing animation, but I can't actually confirm that there was an animation previously. If there is one and it's not just my eyes tricking me, the closing animation would be messed up by this, but I'm pretty sure it doesn't actually have one.

**Review by hmhealey — APPROVED** [review]

File: components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx
Comment (devinbinnie) [inline components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx:357]: Well, hindsight is 20/20 I suppose :P

**Review by devinbinnie — APPROVED** [review]

**Review by jgilliam17 — CHANGES_REQUESTED** [review]
Thanks @lieut-data 
- Characters are rendering as I type; I didn't notice any excessive slowness, although typing/rendering seems to be faster for me on Firefox, than it is on Chrome and Desktop app.
- One issue with the menus - `Move to` sub-menu opens on top of channel menu options
Screen Shot 2021-03-15 at 5 31 46 PM

**Comment by hmhealey** [comment]
I see the issues with the menu positioning as well. It must mess up the position calculation by having no items on first render.

**Comment by lieut-data** [comment]
Oops, ignore this one for now @jgilliam17 -- I clicked on the wrong PR. Still addressing your feedback here.

**Comment by lieut-data** [comment]
Looping in @deanwhillier for the most recent changes, along with a second review requested from @devinbinnie. The delta from the first change rips out `xOffset` in favour of just using `left: 100%` (or `right: 100%` for the plugin post menu context). Across all the browsers I've tested, this seems to "just work", including addressing the issue that Jelena originally found.

To both @hmhealey and my surprise, it doesn't even change the mobile web behaviour, for whatever that's worth :D

**Comment by lieut-data** [comment]
@jgilliam17, I hate to short-circuit QA review before the (second) dev review has completed, but I'm wondering if I can get your eyes on this one in parallel with developer review as it's somewhat high profile and impacting multiple customers. The sidebar positioning /should/ now be resolved :)

**Review by devinbinnie — APPROVED** [review]

**Review by jgilliam17 — APPROVED** [review]
Thank you @lieut-data 
Tested, look good to merge.
- Verified menu issue is fixed. 
- When testing without network or CPU throttle, typing is working as expected. I was able to get some activity on the test server while typing, but it wasn’t enough to make any noticeable change in rendering of the text as I typed.  
- With CPU 4X (and 6X) slowdown I could see some slowness, more so on Ubuntu VM, but I would think that is expected. 
Per our conversation earlier, I'll try using 4x CPU throttle after this is merged to daily and update you on the results.

**Comment by lieut-data** [comment]
Thanks so much, @jgilliam17!

**Comment by amyblais** [comment]
@lieut-data v5.33.1 was just completed, this is ready to be cherry-picked to both v5.33 and v5.31 (based on https://community.mattermost.com/private-core/pl/fjsui3k7yjbx3gqic38k7wu95o).

**Comment by lieut-data** [comment]
Thanks, @amyblais! I'll prepare the backports.

**Comment by amyblais** [comment]
@lieut-data Is your preference to cherry-pick this to ``cloud`` or separately to v5.34 branch once it's created based on March 24th cloud release? (I'm guessing the latter)

**Comment by lieut-data** [comment]
Ahh, hmm. At a minimum, yes to cherry-picking to the `v5.34` branch. I'm 0/5 on whether we push directly to cloud now or wait for it to get picked up organically. Thoughts, @amyblais?

**Comment by amyblais** [comment]
@lieut-data I think cherry-picking to ``cloud`` works too since this has already been soaking on daily for some days. Please help with the ``cloud`` cherry-pick.

**Pull request merged**
